test/test_utils/test_reporting.py

```diff
@@ -0,0 +1,290 @@
+
+        # Regex in order to determine the gpu instance type
+        gpu_instance_pattern = re.compile(r"\w+\.\d*xlarge")
+        gpu_match = gpu_instance_pattern.search(function_keywords)
+
+        if gpu_match:
+            instance_type = gpu_match.group()
+            num_gpus = get_instance_num_gpus(instance_type=instance_type)
+
+            for test in allowed_single_gpu:
+                if test in function_key:
+                    whitelist_single_gpu = True
+                    break
+            if num_gpus == 1:
+                processor = "single_gpu"
+                if not whitelist_single_gpu:
+                    single_gpu_failure_message = (
+                        f"Function uses single-gpu instance type {instance_type}. "
+                        f"Please use multi-gpu instance type."
+                    )
+                    if not self.failure_conditions.get(function_key):
+                        self.failure_conditions[function_key] = [single_gpu_failure_message]
+                    else:
+                        self.failure_conditions[function_key].append(single_gpu_failure_message)
+
+        return processor
+
+    def assemble_report_failure_message(self):
+        """
+        Assemble the failure message if there are any to raise
+
+        :return: the final failure message string
+        """
+        final_message = ""
+        total_issues = 0
+        for func, messages in self.failure_conditions.items():
+            final_message += f"******Problems with {func}:******\n"
+            for idx, message in enumerate(messages):
+                final_message += f"{idx+1}. {message}\n"
+                total_issues += 1
+        final_message += f"TOTAL ISSUES: {total_issues}"
+
+        return final_message, total_issues
+
+    def write_test_coverage_file(self, test_coverage_info):
+        """
+        Write out the test coverage file based on a dictionary defining key/value pairs of test coverage
+        information
+
+        :param test_coverage_info: dict representing the test coverage information
+        """
+        # Assemble the list of headers from one item in the dictionary
+        field_names = []
+        for _key, header in test_coverage_info.items():
+            for field_name, _value in header.items():
+                field_names.append(field_name)
+            break
+
+        # Write to the test coverage file
+        file_open_type = "w+"
+        if self.is_sagemaker and os.path.exists(self.test_coverage_file):
+            file_open_type = "a"
+        with open(self.test_coverage_file, file_open_type) as tc_file:
+            writer = csv.DictWriter(tc_file, delimiter=",", fieldnames=field_names)
+            if file_open_type == "w+":
+                writer.writeheader()
+
+            for _func_key, info in test_coverage_info.items():
+                writer.writerow(info)
+
+    @staticmethod
+    def _infer_field_value(default, options, *comparison_str):
```

Comment:
Why is this a staticmethod?

Reply:
Oops, looks like my editor changed this on me because I didn't need to use any instance attributes/methods in this method. It's not exactly a static method, but also not exactly a general purpose function. I can move this to be just a private function outside of the class